doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,571 @@
+3. **Shared Data Structures**: Use common data structures across passes for consistency
+4. **Lazy Evaluation**: Compute expensive analyses only when needed by downstream passes
+
+## Organizational Structure
+
+### Current Structure
+```
+dace/codegen/
+├── __init__.py
+├── codegen.py              # Main entry point
+├── targets/                # Target-specific generators
+├── instrumentation/        # Profiling support
+├── tools/                  # Utilities
+├── dispatcher.py           # Target routing
+├── control_flow.py         # Control flow analysis
+├── compiler.py             # CMake integration
+├── codeobject.py          # Code file representation
+├── compiled_sdfg.py       # Runtime interface
+└── ...
+```
+
+### Proposed Structure
+```
+dace/codegen/
+├── __init__.py
+├── codegen.py              # Simplified entry point
+├── compiler/               # Compilation and build system
+│   ├── __init__.py
+│   ├── cmake.py            # CMake integration (from compiler.py)
+│   ├── direct.py           # Direct compiler calls
+│   ├── build_system.py     # Abstract build system interface
+│   └── environments.py     # Environment-specific build logic
+├── passes/                 # Code generation passes
+│   ├── __init__.py
+│   ├── analysis/           # Analysis passes
+│   │   ├── __init__.py
+│   │   ├── type_inference.py
+│   │   ├── metadata_collection.py
+│   │   ├── allocation_analysis.py
+│   │   ├── control_flow_analysis.py
+│   │   └── target_analysis.py
+│   ├── transformation/     # Transformation passes
+│   │   ├── __init__.py
+│   │   ├── copy_to_map.py
+│   │   ├── stream_assignment.py
+│   │   ├── tasklet_lowering.py
+│   │   └── memlet_lowering.py
+│   ├── codegen/           # Code generation passes
+│   │   ├── __init__.py
+│   │   ├── state_struct.py
+│   │   ├── allocation_code.py
+│   │   ├── frame_code.py
+│   │   ├── target_code.py
+│   │   └── header_generation.py
+│   ├── file_generation/   # File organization passes
+│   │   ├── __init__.py
+│   │   ├── sdfg_splitting.py
+│   │   └── code_objects.py
+│   └── pipelines.py       # Pre-built pipelines
+├── targets/               # Target-specific generators (simplified)
+│   ├── __init__.py
+│   ├── base.py            # Base target interface (from target.py)
+│   ├── openmp.py          # OpenMP backend (split from cpu.py)
+│   ├── cpp.py             # Pure C++ backend
+│   ├── gpu.py             # GPU backend (generalized from cuda.py)
+│   ├── fpga/              # FPGA backends
+│   └── specialized/       # Other specialized targets
+├── runtime/               # Runtime interface (from compiled_sdfg.py)
+└── utils/                 # Utilities (dispatcher, codeobject, etc.)
+    ├── __init__.py
+    ├── dispatcher.py
+    ├── codeobject.py
+    ├── prettycode.py
+    └── common.py
+```
+
+## Target Refactoring
+
+### Current Issues
+- **CPU backend** actually does OpenMP generation
+- **CUDA backend** is GPU-specific, not general GPU
```

Comment:
I disagree here. I think the targets should be named after the target programming model and not the target hardware. We may indeed want to clarify that the CUDA backend is also used to target HIP/ROCm and refactor accordingly, since they are essentially identical from a user perspective. However, if in the future we want to support other hardware that may also be considered to be a GPU, we probably want to develop completely new targets, e.g., SYCL for Intel GPUs, or Metal for Apple Silicon, etc.